nixos/modules/system/etc/etc.nix

```diff
@@ -250,69 +250,83 @@ in
+          # assume that /etc has already been mounted.
+          # To a large extent this mimics what composefs does. Because
           # it's relatively simple, however, we avoid the composefs dependency.
           # Since this script is not idempotent, it should not run when etc hasn't
           # changed.
           if [[ ! $IN_NIXOS_SYSTEMD_STAGE1 ]] && [[ "${config.system.build.etc}/etc" != "$(readlink -f /run/current-system/etc)" ]]; then
             echo "remounting /etc..."
 
-            tmpMetadataMount=$(mktemp --directory -t nixos-etc-metadata.XXXXXXXXXX)
-            mount --type erofs -o ro ${config.system.build.etcMetadataImage} $tmpMetadataMount
-
-            # Mount the new /etc overlay to a temporary private mount.
-            # This needs the indirection via a private bind mount because you
-            # cannot move shared mounts.
-            tmpEtcMount=$(mktemp --directory -t nixos-etc.XXXXXXXXXX)
-            mount --bind --make-private $tmpEtcMount $tmpEtcMount
-            mount --type overlay overlay \
-              --options lowerdir=$tmpMetadataMount::${config.system.build.etcBasedir},${etcOverlayOptions} \
-              $tmpEtcMount
-
-            # Before moving the new /etc overlay under the old /etc, we have to
-            # move mounts on top of /etc to the new /etc mountpoint.
-            findmnt /etc --submounts --list --noheading --kernel --output TARGET | while read -r mountPoint; do
-              if [[ "$mountPoint" = "/etc" ]]; then
-                continue
-              fi
-
-              tmpMountPoint="$tmpEtcMount/''${mountPoint:5}"
-                ${
-                  if config.system.etc.overlay.mutable then
-                    ''
-                      if [[ -f "$mountPoint" ]]; then
-                        touch "$tmpMountPoint"
-                      elif [[ -d "$mountPoint" ]]; then
-                        mkdir -p "$tmpMountPoint"
-                      fi
-                    ''
-                  else
-                    ''
-                      if [[ ! -e "$tmpMountPoint" ]]; then
-                        echo "Skipping undeclared mountpoint in environment.etc: $mountPoint"
-                        continue
-                      fi
-                    ''
-                }
-              mount --bind "$mountPoint" "$tmpMountPoint"
-            done
-
-            # Move the new temporary /etc mount underneath the current /etc mount.
-            #
-            # This should eventually use util-linux to perform this move beneath,
-            # however, this functionality is not yet in util-linux. See this
-            # tracking issue: https://github.com/util-linux/util-linux/issues/2604
-            ${pkgs.move-mount-beneath}/bin/move-mount --move --beneath $tmpEtcMount /etc
+            # These don't exist yet when we're called by nixos-enter
+            mkdir --parents /.rw-etc/upper /.rw-etc/work
```

Comment:
Wouldn't this always create the `/.rw-etc` folder, even when `system.etc.overlay.mutable = false`? I have all of my systems on read-only `/etc` overlay and the folder doesn't exist there at all, maybe a check could be added so it's only created if `system.etc.overlay.mutable = true` to avoid creating unnecessary folders in the root. Also adding it here means the folder is now created in two places, in addition to inside the initrd: https://github.com/NixOS/nixpkgs/blob/a34c4b8e44635241508126bbd07a1f0219487e5a/nixos/modules/system/etc/etc-activation.nix#L122-L137

Reply:
Good point, I'll have a look at that tomorrow!

Reply:
So, the ones created in initrd are the same ones, since `/sysroot` is what will become the real root once we pivot from initrd into the final system. So there is no issue there. And `mkdir --parents` does nothing if the directories already exist.

I pushed a commit to only create them when using a mutable overlay, so that they are not created for the immutable case.

Reply:
Yep, just thought I would mention they are created twice just in case of future code changes or such. Thank you for the fix! :)